lrpar/src/lib/lex_api.rs

```diff
@@ -3,26 +3,41 @@
 use std::{cmp, error::Error, fmt, hash::Hash, marker};
 
 use cfgrammar::Span;
-use num_traits::{PrimInt, Unsigned};
+use num_traits::{AsPrimitive, PrimInt, Unsigned};
+
+pub trait LexerTypes: fmt::Debug
+where
+    usize: AsPrimitive<Self::StorageT>,
+{
+    type LexemeT: Lexeme<Self::StorageT>;
+    type StorageT: 'static + fmt::Debug + Hash + PrimInt + Unsigned;
+    type LexErrorT: LexError;
+}
 
 /// The base trait which all lexers which want to interact with `lrpar` must implement.
-pub trait Lexer<LexemeT: Lexeme<StorageT>, StorageT: Hash + PrimInt + Unsigned, LexErrorT: LexError>
+pub trait Lexer<
+    StorageT: 'static + Hash + PrimInt + Unsigned,
+    LexerTypesT: LexerTypes<StorageT = StorageT>,
```

Comment:
I need to look at this more carefully, but I couldn't remember why we needed both `StorageT`, and `LexerTypesT::StorageT` here. So I tried to combine them being certain that there was a reason and that would remind me.  The attempt seemed to work, but it is quite possible that I am missing out on some detail here.

https://github.com/ratmice/grmtools/commit/62879fac7422028fdffebcac1177a6a50081ef01

Reply:
I went ahead and tried with `CTLexerBuilder` and `CTParserBuilder` to see if similar change would work there too,
and that seemed to work too.  https://github.com/ratmice/grmtools/commit/db9298ddd0a405140f76b3d8eef567d3882624e1

Reply:
Yes at the moment we don't need a separate `StorageT`. My vague thinking -- and I guess I'll actually have to try this a bit to know if it's a good idea or not! -- is that we might want `ParserTypes` to match `LexerTypes`. An advantage of doing this is that we would open up the possibility of one day having separate types for rules, productions, and so on, without sharing the same `StorageT` -- in a sense, `StorageT` is my hack to avoid proliferating type parameters.

Reply:
Aha, now it makes more sense to me that this is intended as an extension point rather than an opportunity for simplification.

While admittedly not quite understanding the plan in detail one thing I worry about is that `LexerTypes<StorageT = StorageT>` is setting up a type equivalence for these 2 types as a part of the public signature, so there is a tinge of worry that this equivalency "leaks out" in a way that could make changing either of the types be a breaking change.

So there is a bit of hesitation that this could end up in a situation where we could ultimately end up with 2 copies of the same type parameter which must be equivalent, for no gain of flexibility. But I don't feel like I have a good enough understanding here to say whether this skepticism is actually warranted or not.

Reply:
On reflection, I think I was reaching too far, too soon (and, possibly, too far ever!). I've cherry picked your commits and added a minor fix I found when porting something using grmtools.

I have made an attempt to remove more `StorageT`s, but have hit a brick wall where `where` clauses want to proliferate. I'll have to start from scratch, but that can wait until tomorrow. I thought it's useful to let you know that I am still chipping away at this in the meantime!

Reply:
I have spent some time playing around, and I *think* we can do better but it's going to be a lot of work. I might still try doing that, but it's going to be too much for this PR -- and I think this PR does improve the current situation. So I think we should review this PR as-is, consider it for merging, then I may continue experimenting separately.

[Current problems include the fact that the `Step` trait is marked unstable as it's been "redesigned" -- but the redesign happened in 2017... So we can't do `StorageT::zero()..<blah>` which is frustrating.]

Reply:
I think I see what you mean about the `Step` trait, but I've never actually noticed it before. I wonder if using the [`NumOps`](https://docs.rs/num-traits/latest/num_traits/trait.NumOps.html) instead might be an option (this question may be absurd?).

As it is on the PR i'll have another look through, I kind of view this sort of change as a tradeoff.  It kind of makes the implementation more complex, with more difficult trait bounds and such, but avoids pushing lots of generics onto callers.
I also kind of lean towards thinking it seems like an improvement.

Reply:
> I wonder if using the [NumOps](https://docs.rs/num-traits/latest/num_traits/trait.NumOps.html) instead might be an option (this question may be absurd?).

Good question... dunno?!

> It kind of makes the implementation more complex, with more difficult trait bounds and such, but avoids pushing lots of generics onto callers.

Agreed, for 90% of users, it simplifies things. For advanced users, it complicates things, but that's probably a tolerable trade-off.

Reply:
> I have spent some time playing around, and I _think_ we can do better but it's going to be a lot of work. I might still try doing that, but it's going to be too much for this PR -- and I think this PR does improve the current situation. So I think we should review this PR as-is, consider it for merging, then I may continue experimenting separately.

I can agree with this, I went ahead and tried looking into the `StorageT` generics that seemed to be left in `LexerDef` and `LRNonStreamingLexer`, akin to the patches I had previously done in this series, and there is quite a bit of fallout from those.  Is that the kind of stuff you were referring to for a separate PR?

Here is that additional patch https://github.com/ratmice/grmtools/commit/d604ca721ade1fb8f015548eb202cf1efa6d5fc4

Reply:
I was trying that + some other deeper changes, but I should have stopped earlier! I've cherry picked your commit into this branch.